src/dattobd.c

```diff
@@ -4121,13 +4121,95 @@ static int ioctl_dattobd_info(struct dattobd_info *info){
 	return ret;
 }
 
+static int ioctl_dattobd_active_device_info(struct dattobd_active_device_info **adinfo, unsigned int *copyback)
+{
+    /* get the count from *info, free the memory, figure out how much we're actually going to return
+     * reallocate *info to that size, then copy the structs in, set the count for what we're returning
+     * and we're done. */
+	int ret;
+	int active;
+	int user_max; // the count of how many structs the caller has reserved space for
+	int lp;
+	int pos;
+    unsigned long bufsz;
+    struct snap_device *dev;
+    struct dattobd_info *infoitem;
+
+	user_max = (*adinfo)->count;
+	LOG_DEBUG("received dattobd all device info ioctl - max return count %d", user_max);
+
+	// count how many devices there are to return
+	active = 0;
+	for (lp = 0; lp < dattobd_max_snap_devices; lp++)
+	{
+		ret = !snap_devices[lp]; // if I call verify_minor, it logs an error if not in use.
+		if (ret)
+			continue;
+	  	active++;
+	}
+	if (active == 0)
+	{
+		// use existing buffer to send back a zero
+		(*adinfo)->count = 0;
+		*copyback = sizeof(struct dattobd_active_device_info); // tell caller to copy the count back
+	    LOG_DEBUG("No devices active.");
+		return 0;
+	}
+
+	kfree(*adinfo); // free up the buffer allocated in giant switch
+	*adinfo = NULL;
+
+	if (active > user_max)
```

Comment:
I would change the way this works. If the user didnt allocate enough space, the ioctl should return `-ENOSPC` and set the count to the current active count.

Also one line if statements :)